### DISCUSSION TITLE
Way to wait until all tasks are handled

### DISCUSSION THREAD

**Discussion by rafalkrupinski** [taskiq-python/taskiq#350] [Q&A] [answered]
I'm writing a small script that does a lot IO with an async library, so I run it via taskiq to parallelize it, and it works great!

My main task fetches pages of data in a loop and starts new tasks to process, but it doesn't wait for them.

How to detect all the tasks are done to shutdown the broker and safely end the script? I can't just await the main task, as it will finish before other tasks.

**Comment by rafalkrupinski** [comment]
```py
class FinishEventMiddleware(TaskiqMiddleware):
    def __init__(self):
        super().__init__()

        self._count = 0
        self.done = asyncio.Event()

    def pre_execute(self, message: "TaskiqMessage") -> "Union[TaskiqMessage, Coroutine[Any, Any, TaskiqMessage]]":
        if self.done.is_set():
            raise ValueError('Broker is done')
        self._count += 1
        return message

    def post_execute(self, message: "TaskiqMessage", result: "TaskiqResult[Any]") -> None:
        self._count -= 1
        if not self._count:
            self.done.set()
``

**Reply by s3rius** [reply]
I would say that you can try `gather` function `from taksiq import gather`. It will be checking on completion of all tasks in a loop. But this method seems to be working as well.

**Reply by rafalkrupinski** [reply]
actually it can kill the worker between tasks when one task schedules another. Thanks for the answer

**Reply by s3rius** [reply, marked as answer]
@rafalkrupinski, here's a small update. Now when you shutdown worker process it will first stop accepting new tasks and will wait until all tasks are completed. 

https://github.com/taskiq-python/taskiq/releases/tag/0.11.9

Here's doc about it: https://taskiq-python.github.io/guide/cli.html#graceful-and-force-shutdowns
